modules/core/src/opengl.cpp

```diff
@@ -1635,94 +1641,148 @@ Context& initializeContextFromGL()
-            }
+            status = clGetDeviceIDs(platforms[i], CL_DEVICE_TYPE_GPU, devCnt, devices, NULL);
             if (status != CL_SUCCESS)
-                CV_Error_(cv::Error::OpenCLInitError, ("OpenCL: Can't get platform extension string: %d", status));
-
-            if (!strstr((const char*)extensionStr.data(), "cl_khr_gl_sharing"))
-                continue;
-        }
-
-        clGetGLContextInfoKHR_fn clGetGLContextInfoKHR = (clGetGLContextInfoKHR_fn)
-                clGetExtensionFunctionAddressForPlatform(platforms[i], "clGetGLContextInfoKHR");
-        if (!clGetGLContextInfoKHR)
+                CV_Error_(cv::Error::OpenCLInitError, ("OpenCL: Can't get platform devices: %d", status));
+
+            for (unsigned int j = 0; (!sharingSupported && (j < devCnt)); ++j) {
+                size_t extensionSize;
+                status = clGetDeviceInfo(devices[j], CL_DEVICE_EXTENSIONS, 0, NULL, &extensionSize );
+                if (status != CL_SUCCESS)
+                    CV_Error_(cv::Error::OpenCLInitError, ("OpenCL: No devices available: %d", status));
+
+                if(extensionSize > 0)
+                {
+                    char* extensions = nullptr;
+
+                    try {
+                        extensions = new char[extensionSize];
+
+                        status = clGetDeviceInfo(devices[j], CL_DEVICE_EXTENSIONS, extensionSize, extensions, &extensionSize);
+                        if (status != CL_SUCCESS)
+                            continue;
+                    } catch(...) {
+                        CV_Error(cv::Error::OpenCLInitError, "OpenCL: Exception thrown during device extensions gathering");
+                    }
+
+                    std::string devString;
+
+                    if(extensions != nullptr) {
+                        devString = extensions;
+                        delete[] extensions;
+                    }
+                    else {
+                        CV_Error(cv::Error::OpenCLInitError, "OpenCL: Unexpected error during device extensions gathering");
+                    }
+
+                    size_t oldPos = 0;
+                    size_t spacePos = devString.find(' ', oldPos); // extensions string is space delimited
+                    while (spacePos != devString.npos) {
+                        if (strcmp(GL_SHARING_EXTENSION,
+                                devString.substr(oldPos, spacePos - oldPos).c_str())
+                                == 0) {
+                            // Device supports context sharing with OpenGL
+                            devUsed = i;
+                            sharingSupported = true;
+                            break;
+                        }
+                        do {
+                            oldPos = spacePos + 1;
+                            spacePos = devString.find(' ', oldPos);
+                        } while (spacePos == oldPos);
+                    }
+                }
+            }
```

Comment:
Let's extract extension search as dedicated static function. It should make the code more readable.